ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,517 @@
+            .collect::<[Fp61BitPrime; LargeProofGenerator::PROOF_LENGTH]>();
+        // consume the rest
+        let proofs = iterator.collect::<Vec<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>>();
+        ProofBatch {
+            first_proof,
+            proofs,
+        }
+    }
+}
+
+impl ProofBatch {
+    /// This function returns the length in field elements.
+    fn len(&self) -> usize {
+        self.proofs.len() * SmallProofGenerator::PROOF_LENGTH + LargeProofGenerator::PROOF_LENGTH
+    }
+
+    /// This function returns an iterator over the field elements of all proofs.
+    fn iter(&self) -> impl Iterator<Item = &Fp61BitPrime> {
+        self.first_proof
+            .iter()
+            .chain(self.proofs.iter().flat_map(|x| x.iter()))
+    }
+
+    /// This function sends a `Proof` to the party on the left
+    async fn send_to_left<C>(&self, ctx: &C) -> Result<(), Error>
+    where
+        C: Context,
+    {
+        // set up context
+        let ctx_left = ctx.set_total_records(self.len());
+
+        // set up channel
+        let send_channel_left =
+            &ctx_left.send_channel::<Fp61BitPrime>(ctx.role().peer(Direction::Left));
+
+        // send to left
+        // we send the proof batch via sending the individual field elements
+        ctx_left
+            .parallel_join(
+                self.iter().enumerate().map(|(i, x)| async move {
+                    send_channel_left.send(RecordId::from(i), x).await
+                }),
+            )
+            .await?;
+        Ok(())
+    }
+
+    /// This function receives a `Proof` from the party on the right.
+    async fn receive_from_right<C>(ctx: &C, length: usize) -> Result<Self, Error>
+    where
+        C: Context,
+    {
+        // set up context
+        let ctx_left = ctx.set_total_records(length);
+
+        // set up channel
+        let receive_channel_right =
+            &ctx_left.recv_channel::<Fp61BitPrime>(ctx.role().peer(Direction::Right));
+
+        // receive from the right
+        Ok(ctx
+            .parallel_join(
+                (0..length)
+                    .map(|i| async move { receive_channel_right.receive(RecordId::from(i)).await }),
+            )
+            .await?
+            .into_iter()
+            .collect())
+    }
+
+    /// This function generates a batch of quadratic proofs
+    /// it iteratively generates proofs and collects them in a batch
+    ///
+    /// It generates `batch_from_left` as its left output
+    /// and `prover_left_batch` as its right output.
+    /// Further, it outputs the generated masks needed for the verification of proof from the left and right.
+    ///
+    /// Outputs `(batch_from_left, prover_left_batch, (p_mask, q_mask))`
+    ///
+    /// ## Panics
+    /// Panics when the recursion factor is smaller than 2 or masks cannot be set safely (implementation issue).
+    fn compute_proof<C, I>(ctx: &C, uv_tuple_block: I) -> (Self, Self, (Fp61BitPrime, Fp61BitPrime))
```

Comment:
We can't have `ProofBatch::compute_proof`, `ProofGenerator::compute_proof` and `ProofGenerator::compute_next_proof`.

I think a lot of the logic can move into "prover.rs" and "verifier.rs". If we still feel it is necessary to have another class that does the whole "batch" thing, maybe we need a new name for it (not ProofBatch) and the function should have a more descriptive and unique name.

Reply:
Also, this function is way too long. I'm surprised clippy didn't complain about it. We need to shorten this function.

Reply:
as mentioned above, the proof batch uses proofs from two different proof generators.